src/thinblock.h

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2016 The Bitcoin Unlimited developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_THINBLOCK_H
+#define BITCOIN_THINBLOCK_H
+
+#include "serialize.h"
+#include "uint256.h"
+#include "primitives/block.h"
+#include "bloom.h"
+
+#include <vector>
+
+class CThinBlock
+{
+public:
+    CBlockHeader header;
+    std::vector<uint256> vTxHashes; // List of all transactions id's in the block
+    std::map<uint256, CTransaction> mapMissingTx; // map of transactions that did not match the bloom filter
+
+public:
+
+    /**
+     * Create from a CThinBlock, finding missing transactions according to filter
+     */
+    CThinBlock(const CBlock& block, CBloomFilter& filter);
+
+    CThinBlock() {}
+
+    int64_t GetBlockTime() { return header.GetBlockTime(); }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(header);
+        READWRITE(vTxHashes);
+        READWRITE(mapMissingTx);
+
+    }
+};
+
+class CXThinBlock
+{
+public:
+    CBlockHeader header;
+    std::vector<uint64_t> vTxHashes; // List of all transactions id's in the block
+    std::map<uint256, CTransaction> mapMissingTx; // map of transactions that did not match the bloom filter
+    bool collision;
+
+public:
+
+    /**
+     * Create from a CThinBlock, finding missing transactions according to filter
+     */
+    CXThinBlock(const CBlock& block, CBloomFilter& filter);
+
+    CXThinBlock() {}
+
+    int64_t GetBlockTime() { return header.GetBlockTime(); }
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action, int nType, int nVersion) {
+        READWRITE(header);
+        READWRITE(vTxHashes);
+        READWRITE(mapMissingTx);
+
+    }
+};
+
+// This class is used for retrieving a list of still missing transactions after receiving a "thinblock" message.
+// The CXThinBlockTx when recieved can be used to fill in the missing transactions after which it is sent
+// back to the requestor.  This class uses a 64bit hash as opposed to the normal 256bit hash.
+class CXThinBlockTx
```

Comment:
would it be possible to use different lengths of the hash and send the length in the thin block?  I haven't done the math but I'll bet that the probability of collision is very low with 32 bits and a few thousand txns...